File: src/main/java/org/springframework/samples/nt4h/card/product/inGame/ProductInGame.java
package org.springframework.samples.nt4h.card.product.inGame;

import lombok.*;
import org.springframework.samples.nt4h.card.product.Product;
import org.springframework.samples.nt4h.card.product.StateProduct;
import org.springframework.samples.nt4h.game.Game;
import org.springframework.samples.nt4h.model.NamedEntity;
import org.springframework.samples.nt4h.player.Player;

import javax.persistence.*;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

@Entity
@Getter
@Setter
@Table(name = "products_in_game")
public class ProductInGame extends NamedEntity {

    public ProductInGame(Product product,Game game) {
        this.product = product;
        this.stateProduct = StateProduct.IN_SALE;
        this.game = game;
        setName(product.getName());
    }

    @NotNull
    @Enumerated(EnumType.STRING)
    private StateProduct stateProduct;

    @ManyToOne(cascade = CascadeType.ALL)
    private Product product;

    @ManyToOne(cascade = CascadeType.ALL)
    private Player player;

    @NotNull
    @ManyToOne(cascade = CascadeType.ALL)
    private Game game;

    public void onDeleteSetNull() {
        product = null;
        player = null;
        game = null;
    }
}

File: src/main/kotlin/org/springframework/samples/nt4h/card/product/ProductInGame.kt
package org.springframework.samples.nt4h.card.product

import org.springframework.samples.nt4h.card.TimesUsed
import org.springframework.samples.nt4h.game.Game
import org.springframework.samples.nt4h.model.NamedEntity
import org.springframework.samples.nt4h.player.Player
import javax.persistence.CascadeType
import javax.persistence.Entity
import javax.persistence.Enumerated
import javax.persistence.ManyToOne
import javax.validation.constraints.Positive

@Entity
data class ProductInGame(
    // Attributes -------------------------------------------------------------
    @Positive
    var timesUsed: Int = 0,
    @Enumerated
    var stateProduct: StateProduct = StateProduct.UNDEFINED,
    // var quantityRemaining: Int = 0, // TODO: Cambiar por un enumerado
    var isActive: Boolean = true,

    // Relationships ----------------------------------------------------------
    @ManyToOne(cascade = [CascadeType.ALL])
    var product: Product? = null,
    @ManyToOne(cascade = [CascadeType.ALL])
    var player: Player? = null
) : NamedEntity() {

    fun isAvailable(): Boolean {
        return product?.limit?.haveMoreUses(timesUsed) ?: false
    }
}
